File: Sources/CountriesFlagsObjC/source/Countries/UIImage+TimorLeste.m
//
//  UIImage+TimorLeste.m
//  CountriesFlags
//

#import "Countries/UIImage+TimorLeste.h"

#import "Extensions/UIBezierPath+Triangle.h"
#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (TimorLeste)

+ (UIImage * _Nullable)flagTimorLesteWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:218.0f/255.0f
                                        green:41.0f/255.0f
                                         blue:28.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorBlack = [UIColor blackColor];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:199.0f/255.0f
                                            blue:44.0f/255.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    [[UIBezierPath triangleInSize:size type:FlagTriangleTypeLeftMiddle] fill];
    
    [colorBlack setFill];
    [[UIBezierPath triangleInSize:CGSizeMake(size.width / 36.0f * 24.0f, size.height)
                             type:FlagTriangleTypeLeftMiddle] fill];
    
    [colorWhite setFill];
    CGPoint starCenter = CGPointMake(size.width / 36.0f * 4.0f, size.height / 2.0f);
    CGFloat radius = size.height / 18.0 * 6.0 / 2.0;
    UIBezierPath* path = [UIBezierPath starAtPoint:starCenter radius:radius type:FlagStarTypeFivePointed];
    [path rotateToAngle:-atan(4.0 / 9.0) * 180.0 / M_PI center:starCenter];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+TimorLeste.swift
//
//  UIImage+TimorLeste.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Timor-Leste "wikipedia"
     
     Generates the flag of Timor-Leste.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Timor-Leste.
     */
    internal static func flagTimorLeste(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 218.0 / 255.0,
                               green: 41.0 / 255.0,
                               blue: 28.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white
        let colorBlack = UIColor.black
        let colorYellow = UIColor(red: 1.0,
                                  green: 199.0 / 255.0,
                                  blue: 44.0 / 255.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIBezierPath.triangle(in: size, type: .leftMiddle).fill()

        colorBlack.setFill()
        UIBezierPath.triangle(in: CGSize(width: size.width / 36.0 * 24.0, height: size.height),
                              type: .leftMiddle).fill()

        colorWhite.setFill()
        let radius = CGFloat(size.height / 18.0 * 6.0) / 2.0
        let path = UIBezierPath.star(with: CGSize(width: size.width / 36.0 * 8.0, height: size.height),
                                     radius: radius)
        path.rotate(to: -atan(4.0 / 9.0) * 180.0 / .pi,
                    around: CGPoint(x: size.width / 36.0 * 8.0 / 2.0, y: size.height / 2.0))
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
